src/upgrade/upgradeManager.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { CodeActionKind, commands, type ExtensionContext, languages, workspace, type WorkspaceFolder } from "vscode";
+import * as semver from 'semver'
+import { Jdtls } from "../java/jdtls";
+import { languageServerApiManager } from "../languageServerApi/languageServerApiManager";
+import { NodeKind, type INodeData } from "../java/nodeData";
+import issueManager from "./issueManager";
+import { Upgrade } from "../constants";
+import { UpgradeReason } from "./type";
+import { instrumentOperationAsVsCodeCommand } from "vscode-extension-telemetry-wrapper";
+import { Commands } from "../commands";
+import metadataManager from "./metadataManager";
+import UpgradeCodeActionProvider from "./upgradeCodeActionProvider";
+import UpgradeCodeLensProvider from "./upgradeCodeLensProvider";
+import { buildPackageId } from "./utility";
+import pomDataManager from "./pomDataManager";
+
+const DEFAULT_UPGRADE_PROMPT = "Upgrade Java project dependency";
+
+class UpgradeManager {
+    public initialize(context: ExtensionContext) {
+        // Command to be used
+        context.subscriptions.push(instrumentOperationAsVsCodeCommand(Commands.VIEW_TRIGGER_JAVA_UPGRADE_TOOL, (promptText?: string) => {
+            this.runUpgrade(promptText ?? DEFAULT_UPGRADE_PROMPT);
+        }));
+
+        // Quick Fix provider
+        context.subscriptions.push(languages.registerCodeActionsProvider(
+            {
+                language: "xml",
+                pattern: "**/pom.xml"
+            },
+            new UpgradeCodeActionProvider(),
+            {
+                providedCodeActionKinds: [CodeActionKind.QuickFix],
+            }
+        ));
+        context.subscriptions.push(languages.registerCodeLensProvider(
+            {
+                language: "xml",
+                pattern: "**/pom.xml"
+            },
+            new UpgradeCodeLensProvider(),
+        ));
+
+        // Metadata update & initial scan
+        metadataManager.tryRefreshMetadata(context).then(() => {
+            upgradeManager.scan();
+        });
+    }
+
+    public scan() {
+        workspace.workspaceFolders?.forEach((folder) =>
+            this.checkUpgradableComponents(folder)
+        );
+    }
+
+    private async checkUpgradableComponents(folder: WorkspaceFolder) {
```

Comment:
Since we are already in Project Explorer, we already know all the dependencies and the version of java runtimes. Can we work with the project metadata directly? wihtout parsing the POM/Gradle files manually? CC @jdneo